Add basic support for inline javascript

smathot opened the pull request:
Allows for inline JavaScript code in OSWeb through the `inline_javascript` plugin that is bundled with the OSWeb extension. To allow JavaScript to be executed on the desktop, [Js2Py](https://github.com/PiotrDabkowski/Js2Py) is used.

Supports:

- Access to the `vars` object for setting experimental variables.
- ECMA Script 5.1 within Python. There is some flexibility here. For example, the `let` keyword is accepted even though this is ECMA Script 6.

Limitations: 

- Does not support `Canvas` objects and other parts of the Python API. This would be possible, but would require considerable improvements to the JavaScript implementation of these classes.
- <strike>When there is an error in the JavaScript code and the experiment is executed in OpenSesame, the process closes silently, without any error message. This strikes me as a bug/ limitation in js2py.</strike>

Several packages have been updated to support this:

- shyras/osweb@6ba2cf6 (see shyras/osweb#121)
- smathot/OpenSesame@53d491e
- smathot/QProgEdit@f7c5fa7   (`release/4.1.0`)
- smathot/opensesame-extension-osweb@d4bab0a (see smathot/opensesame-extension-osweb#12)

smathot committed:
Add basic support for inline javascript
- Bump to 1.3.0

smathot committed:
Add basic support for inline javascript
- Bump to 1.3.0

smathot commented on src/js/osweb/items/inline_javascript.js:
```diff
@@ -0,0 +1,92 @@
+     * Parse a definition string and retrieve all properties of the item.
+     * @param {String} script - The script containing the properties of the item.
+     */
+  from_string (script) {
```
@shyras I'm not so happy with the implementation of `from_script()`, because it treats the `_run` and `_prepare` variables as though they are special cases, which they arent: every item should be able to parse multiline variables of any name.

shyras replied:
Ok. That is something which we should look into. I thought that it were special cases. So far as I can see looking at the code, Daniel made some changes to the routine. Maybe he knows a quick fix, otherwise we should place it on the to-do list.

dschreij replied:
But this is a whole new class, right? The one that Sebastiaan just added. The `from_script()` function is not here, or do you mean the `from_string()` function?

smathot replied:
Sorry, I mean the `from_string()` function indeed. I copied that from `inline_script.js`. It's not important for now, but the multiline-variable parsing should ideally be carried into the general functionality for parsing variables, which I assume happens in `item.js`?

dschreij commented:
If you want to support any version of javascript, it may be worth having a look at https://babeljs.io/. This is an (industry standard) command line tool that transpiles newer syntax down to any version of JavaScript. This would allow the user to use the newest JS syntax, and when saving, OpenSesame can pipe the script to babel, which will return it in a version that is always compatible with js2py.

I don't think this extra step generates a lot of overhead, but it will provide *a lot* more convenience, as the older versions of JavaScript tend to be very verbose and limited in what the standard library offers regarding functionality. I'd have at least have a look at the possibilities of Babel.

smathot commented:
BabelJS looks nice, but if I understand correctly it requires a full nodejs installation, right?

dschreij commented:
Not really. There are solutions that work standalone and are just a single binary to which you can pass code or files, but I can't currently find them. I have only been able to do a quick search now, and these options all still use npm to set things up:
https://github.com/babel/babel/tree/master/packages/babel-standalone
https://babeljs.io/docs/en/babel-cli

This might be a good all-in-one solution: It is a python library which already does something similar as js2py and has Babel transpilation built in: https://github.com/amol-/dukpy

dschreij commented:
By the way, js2py also appears to have ES6 support according to its pypi page: https://pypi.org/project/Js2Py/
There should be a function `js2py.eval_js6()`

smathot commented:
Ok, but let's put the ES6 issue aside for now though. (Unless there is a very easy solution, but not as something to focus on initially.) Most users will just do basic scripting for which this isn't crucial. It will be more important to develop an API that allows users interact with the `Canvas`, `Keyboard` etc. Ideally in a way that is virtually identical to the Python API.

smathot commented:
Two minor updates:

- The issue with the experiment process being killed when there was an issue in the JavaScript is resolved. This actually had to do with the js2py `Exception` options not being picklable and therefore not being transferred to the GUI.
- js2py does indeed support (experimantally) ES6, but it appears they also do this through Babel. Actually they have a translated (!) implementation of Babel, but they warn against using it because it's really big and slow.

dschreij commented:
Sorry, I still didn't get around to testing this due to a swamped schedule and life in general. I hope to make some time in the next 2-3 days. This is not something you're looking to release quite soon, are you?
Have you been able to look into `morph-expressions` for js compiling?

smathot committed:
Merge branch 'inline_javascript' of github.com:smathot/osweb into inline_javascript

dschreij commented on example-experiments/inline_javascript.osexp:
```diff
@@ -0,0 +1,149 @@
+define sketchpad instructions
+	set duration keypress
+	set description "Displays stimuli"
+	draw textline center=1 color=white font_bold=no font_family=mono font_italic=no font_size=18 html=yes show_if=always text="INSTRUCTIONS<br /><br />Press 'z' if you see a blue F<br />Press 'z' if you don't<br /><br />Press any key to continue!" x=0 y=0 z_index=0
```
This should be press 'm' if you don't

dschreij reviewed (approved):
Looks good, and works flawlessly for me. I only found a small error in the instructions of the example experiment, as indicated.
I do recommend to not use 'eval()' directly, as it is generally advised against by the js community due to security reasons. You could use `morph-expressions` to precompile js code (as described earlier), which should generally be safer and maybe even more performant.

smathot committed:
JavaScript workspace: add proxy for var_store

smathot committed:
VarStore: improvements for inline_javascript
- respect `evaluate` argument
- add `addQuotes` argument
- set property to bypass proxy

smathot committed:
JavaScript workspace: various improvements
- add as experiment property
- fix Proxy

smathot commented:
I did some more work on this. The Proxy turned out more interesting than I envisioned. Apparently, when Proxied, property calls in the `VarStore` itself are also captured by the Proxy. This resulted in a feedback loop that made it impossible to *not* evaluate variables.

In addition, there were a few bugs/ oddities:

- The `evaluate` argument to `VarStore.get()` was simply ignored
- `VarStore.get()` quotes strings. The reason for this is that strings should be quoted in the context of conditional statements. However, this should be implemented differently. I suggest that we leave this for a larger refactoring that I will open an issue for later. For now I simply added an `addQuotes` argument to at least disable this in the context of inline scripts.

@dschreij Could you take a look at the new commits?

dschreij reviewed (approved):
Everything appears to work (really) well. The only thing is that the inline javascript experiment logic is erroneous: the correct response mapping is often incorrect (z is randomly counted as correct for both trials in which a blue z is present or absent), and the first instruction slide tells to press z for both present and absent trials. But this is an example or test experiment anyway, and I don't see any other problems on the JavaScript side.

smathot commented:
Ow that's right. You fixed that in your PR, but I closed that. I'll fix that just to have things be a little neat.
